massa-bootstrap/src/server.rs

```diff
@@ -14,363 +42,451 @@ use massa_signature::KeyPair;
+        self.listen_handle.abort();
+        self.update_handle.abort();
+        self.main_handle.join().unwrap()
     }
 }
 
-#[allow(clippy::too_many_arguments)]
-/// TODO merging the command senders into one channel structure may allow removing that allow
-///
-/// start a bootstrap server.
-/// Once your node will be ready, you may want other to bootstrap from you.
+/// See module level documentation for details
 pub async fn start_bootstrap_server(
     consensus_controller: Box<dyn ConsensusController>,
     network_command_sender: NetworkCommandSender,
     final_state: Arc<RwLock<FinalState>>,
-    bootstrap_config: BootstrapConfig,
-    establisher: Establisher,
+    config: BootstrapConfig,
+    mut establisher: Establisher,
     keypair: KeyPair,
     version: Version,
-) -> Result<Option<BootstrapManager>, BootstrapError> {
+) -> Result<Option<BootstrapManager>, Box<BootstrapError>> {
     massa_trace!("bootstrap.lib.start_bootstrap_server", {});
-    if let Some(bind) = bootstrap_config.bind {
-        let (manager_tx, manager_rx) = mpsc::channel::<()>(1);
+    let Some(listen_addr) = config.listen_addr else {
+        return Ok(None);
+    };
 
-        let join_handle = tokio::spawn(async move {
-            BootstrapServer {
-                consensus_controller,
-                network_command_sender,
-                final_state,
-                establisher,
-                manager_rx,
-                bind,
-                keypair,
-                version,
-                ip_hist_map: HashMap::with_capacity(bootstrap_config.ip_list_max_size),
-                bootstrap_config,
-            }
-            .run()
-            .await
-        });
-        Ok(Some(BootstrapManager {
-            join_handle,
-            manager_tx,
-        }))
-    } else {
-        Ok(None)
-    }
+    // TODO(low prio): See if a zero capacity channel model can work
+    let (stopper_tx, stopper_rx) = crossbeam::channel::bounded::<()>(1);
+
+    let runtime = Runtime::new().expect("Failed to create a bootstrap runtime");
+    let listener = establisher
+        .get_listener(listen_addr)
+        .await
+        .map_err(BootstrapError::IoError)?;
+
+    let Ok(max_bootstraps) = config.max_simultaneous_bootstraps.try_into() else {
+        return Err(BootstrapError::GeneralError("Fail to convert u32 to usize".to_string()).into());
+    };
+    // This is the primary interface between the async-listener, and the (soon to be) sync worker
+    let (listener_tx, listener_rx) =
+        crossbeam::channel::bounded::<(Duplex, SocketAddr)>(max_bootstraps);
+
+    let allow_block_list = SharedAllowBlockList::new(
+        config.bootstrap_whitelist_path.clone(),
+        config.bootstrap_blacklist_path.clone(),
+    )
+    .map_err(BootstrapError::GeneralError)?;
+
+    let update_handle = runtime.handle().clone().spawn(BootstrapServer::run_updater(
+        allow_block_list.clone(),
+        config.cache_duration.into(),
+    ));
```

Comment:
This is where I'm a bit hazey about how to best use a created handle. Will exist on a thread one it's non-`async`, but will need to add-in some sort of stop-control. Perhaps an `Arc<AtomicBool>` that can break the loop and that would allow for a join on the thread.

Reply:
I think I'm getting a better handle on things now.

key-word: _**better**_